File: android/src/main/java/com/getcapacitor/plugin/http/Http.java
package com.getcapacitor.plugin.http;

import android.Manifest;
import android.util.Base64;
import android.util.Log;
import com.getcapacitor.CapConfig;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import java.io.BufferedReader;
import java.io.ByteArrayOutputStream;
import java.io.DataOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpCookie;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * Native HTTP Plugin
 */
@CapacitorPlugin(
    name = "Http",
    permissions = {
        @Permission(strings = { Manifest.permission.WRITE_EXTERNAL_STORAGE }, alias = "HttpWrite"),
        @Permission(strings = { Manifest.permission.WRITE_EXTERNAL_STORAGE }, alias = "HttpRead")
    }
)
public class Http extends Plugin {

    public static final int HTTP_REQUEST_DOWNLOAD_WRITE_PERMISSIONS = 9022;
    public static final int HTTP_REQUEST_UPLOAD_READ_PERMISSIONS = 9023;

    CapConfig capConfig;
    CapacitorCookieManager cookieManager;

    /**
     * Helper function for getting the serverUrl from the Capacitor Config. Returns an empty
     * string if it is invalid and will auto-reject through {@code call}
     * @param call the {@code PluginCall} context
     * @return the string of the server specified in the Capacitor config
     */
    private String getServerUrl(PluginCall call) {
        String url = call.getString("url", "");

        URI uri = getUri(url);
        if (uri == null) {
            call.reject("Invalid URL. Check that \"server\" is passed in correctly");
            return "";
        }

        return url;
    }

    /**
     * Try to parse a url string and if it can't be parsed, return null
     * @param url the url string to try to parse
     * @return a parsed URI
     */
    private URI getUri(String url) {
        try {
            return new URI(url);
        } catch (Exception ex) {
            return null;
        }
    }

    private boolean isStoragePermissionGranted(PluginCall call, String permission) {
        if (hasPermission(permission)) {
            Log.v(getLogTag(), "Permission '" + permission + "' is granted");
            return true;
        } else {
            Log.v(getLogTag(), "Permission '" + permission + "' denied. Asking user for it.");
            requestPermissions(call);
            return false;
        }
    }

    private void http(final PluginCall call, final String httpMethod) {
        new Thread(
            new Runnable() {
                public void run() {
                    try {
                        JSObject response = HttpRequestHandler.request(call, httpMethod);
                        call.resolve(response);
                    } catch (Exception e) {
                        System.out.println(e.toString());
                        call.reject(e.getClass().getSimpleName(), e);
                    }
                }
            }
        )
            .start();
    }

    @Override
    public void load() {
        this.cookieManager = new CapacitorCookieManager(null, java.net.CookiePolicy.ACCEPT_ALL);
        java.net.CookieHandler.setDefault(cookieManager);
        capConfig = getBridge().getConfig();
    }

    @PluginMethod
    public void request(final PluginCall call) {
        this.http(call, null);
    }

    @PluginMethod
    public void get(final PluginCall call) {
        this.http(call, "GET");
    }

    @PluginMethod
    public void post(final PluginCall call) {
        this.http(call, "POST");
    }

    @PluginMethod
    public void put(final PluginCall call) {
        this.http(call, "PUT");
    }

    @PluginMethod
    public void patch(final PluginCall call) {
        this.http(call, "PATCH");
    }

    @PluginMethod
    public void del(final PluginCall call) {
        this.http(call, "DELETE");
    }

    @PluginMethod
    public void downloadFile(PluginCall call) {
        try {
            bridge.saveCall(call);
            String fileDirectory = call.getString("fileDirectory", FilesystemUtils.DIRECTORY_DOCUMENTS);

            if (
                !FilesystemUtils.isPublicDirectory(fileDirectory) ||
                isStoragePermissionGranted(call, Manifest.permission.WRITE_EXTERNAL_STORAGE)
            ) {
                call.release(bridge);
                JSObject response = HttpRequestHandler.downloadFile(call, getContext());

                call.resolve(response);
            }
        } catch (MalformedURLException ex) {
            call.reject("Invalid URL", ex);
        } catch (IOException ex) {
            call.reject("IO Error", ex);
        } catch (Exception ex) {
            call.reject("Error", ex);
        }
    }

    @PluginMethod
    public void uploadFile(PluginCall call) {
        try {
            String fileDirectory = call.getString("fileDirectory", FilesystemUtils.DIRECTORY_DOCUMENTS);
            bridge.saveCall(call);

            if (
                !FilesystemUtils.isPublicDirectory(fileDirectory) ||
                isStoragePermissionGranted(call, Manifest.permission.WRITE_EXTERNAL_STORAGE)
            ) {
                call.release(bridge);
                JSObject response = HttpRequestHandler.uploadFile(call, getContext());
                call.resolve(response);
            }
        } catch (Exception ex) {
            call.reject("Error", ex);
        }
    }

    @PluginMethod
    public void setCookie(PluginCall call) {
        String key = call.getString("key");
        String value = call.getString("value");
        String url = getServerUrl(call);

        if (!url.isEmpty()) {
            cookieManager.setCookie(url, key, value);
            call.resolve();
        }
    }

    @PluginMethod
    public void getCookiesMap(PluginCall call) {
        String url = getServerUrl(call);
        if (!url.isEmpty()) {
            HttpCookie[] cookies = cookieManager.getCookies(url);
            JSObject cookiesJsObject = new JSObject();
            for (HttpCookie cookie : cookies) {
                cookiesJsObject.put(cookie.getName(), cookie.getValue());
            }
            call.resolve(cookiesJsObject);
        }
    }

    @PluginMethod
    public void getCookies(PluginCall call) {
        String url = getServerUrl(call);
        if (!url.isEmpty()) {
            HttpCookie[] cookies = cookieManager.getCookies(url);
            JSArray cookiesJsArray = new JSArray();
            for (HttpCookie cookie : cookies) {
                JSObject cookieJsPair = new JSObject();
                cookieJsPair.put("key", cookie.getName());
                cookieJsPair.put("value", cookie.getValue());
                cookiesJsArray.put(cookieJsPair);
            }
            JSObject cookiesJsObject = new JSObject();
            cookiesJsObject.put("cookies", cookiesJsArray);
            call.resolve(cookiesJsObject);
        }
    }

    @PluginMethod
    public void getCookie(PluginCall call) {
        String key = call.getString("key");
        String url = getServerUrl(call);
        if (!url.isEmpty()) {
            HttpCookie cookie = cookieManager.getCookie(url, key);
            JSObject cookieJsObject = new JSObject();
            cookieJsObject.put("key", key);
            if (cookie != null) {
                cookieJsObject.put("value", cookie.getValue());
            } else {
                cookieJsObject.put("value", "");
            }
            call.resolve(cookieJsObject);
        }
    }

    @PluginMethod
    public void deleteCookie(PluginCall call) {
        String key = call.getString("key");
        String url = getServerUrl(call);
        if (!url.isEmpty()) {
            cookieManager.setCookie(url, key + "=; Expires=Wed, 31 Dec 2000 23:59:59 GMT");
            call.resolve();
        }
    }

    @PluginMethod
    public void clearCookies(PluginCall call) {
        cookieManager.removeAllCookies();
        call.resolve();
    }
}

File: android/src/main/java/com/getcapacitor/plugin/http/Http.kt
package com.getcapacitor.plugin.http

import android.Manifest
import android.util.Log
import com.getcapacitor.*
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.getcapacitor.plugin.http.CapacitorHttpHandler.download
import com.getcapacitor.plugin.http.CapacitorHttpHandler.request
import com.getcapacitor.plugin.http.CapacitorHttpHandler.upload
import com.getcapacitor.plugin.http.FileExtensions.isPublicDirectory
import com.getcapacitor.plugin.http.cookie.CapacitorCookieManager
import java.lang.Exception
import java.net.URI

/**
 * Native HTTP Plugin
 */
@CapacitorPlugin(
    name = "Http",
    permissions = [
        Permission(strings = [Manifest.permission.WRITE_EXTERNAL_STORAGE], alias = "HttpWrite"),
        Permission(strings = [Manifest.permission.WRITE_EXTERNAL_STORAGE], alias = "HttpRead")
    ]
)
class Http : Plugin() {
    private lateinit var capConfig: CapConfig
    private lateinit var cookieManager: CapacitorCookieManager

    override fun load() {
        capConfig = getBridge().config
        cookieManager = CapacitorCookieManager(context)
    }

    /**
     * Helper function for getting the serverUrl from the Capacitor Config. Returns an empty
     * string if it is invalid and will auto-reject through `call`
     * @param call the `PluginCall` context
     * @return the string of the server specified in the Capacitor config
     */
    private fun getServerUrl(call: PluginCall): URI? {
        val url = call.getString("url", "http://localhost")
        return getUri(url)
    }

    /**
     * Try to parse a url string and if it can't be parsed, return null
     * @param url the url string to try to parse
     * @return a parsed URI
     */
    private fun getUri(url: String?): URI? {
        return try { URI(url) } catch (ex: Exception) { null }
    }

    private fun isStoragePermissionGranted(call: PluginCall, permission: String): Boolean {
        return if (hasPermission(permission)) {
            Log.v(logTag, "Permission '$permission' is granted")
            true
        } else {
            Log.v(logTag, "Permission '$permission' denied. Asking user for it.")
            requestPermissions(call)
            false
        }
    }

    private fun http(call: PluginCall, httpMethod: String) {
        val asyncHttpCall = Runnable {
            try {
                val response = request(call, httpMethod, cookieManager)
                call.resolve(response)
            } catch (e: Exception) {
                println(e.toString())
                call.reject(e.javaClass.simpleName, e)
            }
        }
        val httpThread = Thread(asyncHttpCall)
        httpThread.start()
    }

    @PluginMethod
    fun request(call: PluginCall) {
        val method = call.getString("method") ?: return call.reject("When using the \"request\" function, a \"method\" parameter is required.")
        http(call, method)
    }

    @PluginMethod
    fun get(call: PluginCall) {
        http(call, "GET")
    }

    @PluginMethod
    fun post(call: PluginCall) {
        http(call, "POST")
    }

    @PluginMethod
    fun put(call: PluginCall) {
        http(call, "PUT")
    }

    @PluginMethod
    fun patch(call: PluginCall) {
        http(call, "PATCH")
    }

    @PluginMethod
    fun del(call: PluginCall) {
        http(call, "DELETE")
    }

    @PluginMethod
    fun downloadFile(call: PluginCall) {
        try {
            bridge.saveCall(call)
            val fileDirectory = call.getString("fileDirectory", FileExtensions.DIRECTORY_DOCUMENTS)
            if (!isPublicDirectory(fileDirectory!!) ||
                    isStoragePermissionGranted(call, Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
                call.release(bridge)
                var callback: IProgressCallback? = object : IProgressCallback {
                    override fun onProgress(progress: Double) {

                    }
                }
                val isProgressTracked = call.getBoolean("progress", false)
                if (isProgressTracked!!) {
                    callback = object : IProgressCallback {
                        override fun onProgress(progress: Double) {
                            val ret = JSObject()
                            ret.put("type", "DOWNLOAD")
                            ret.put("url", call.getString("url"))
                            ret.put("bytes", progress)
                            notifyListeners("progress", ret)
                        }
                    }
                }
                val response = download(call, context, cookieManager, callback!!)
                call.resolve(response)
            }
        } catch (ex: Exception) {
            call.reject("Error", ex)
        }
    }

    @PluginMethod
    fun uploadFile(call: PluginCall) {
        try {
            val fileDirectory = call.getString("fileDirectory", FileExtensions.DIRECTORY_DOCUMENTS)
            bridge.saveCall(call)
            if (!isPublicDirectory(fileDirectory!!) ||
                    isStoragePermissionGranted(call, Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
                call.release(bridge)
                val response = upload(call, context, cookieManager)
                call.resolve(response)
            }
        } catch (ex: Exception) {
            call.reject("Error", ex)
        }
    }

    @PluginMethod
    fun setCookie(call: PluginCall) {
        val key = call.getString("key") ?: return call.reject("\"key\" must not be null")
        val value = call.getString("value") ?: return call.reject("\"value\" must not be null")
        val url = getServerUrl(call) ?: return call.reject("Could not parse the server \"url\" value")

        cookieManager.setCookie(url, key, value)
        call.resolve()
    }

    @PluginMethod
    fun getCookiesMap(call: PluginCall) {
        val cookies = cookieManager.getCookies()
        val cookiesJsObject = JSObject()
        for (cookie in cookies) {
            cookiesJsObject.put(cookie.name, cookie.value)
        }
        call.resolve(cookiesJsObject)
    }

    @PluginMethod
    fun getCookies(call: PluginCall) {
        val cookies = cookieManager.getCookies()
        val cookiesJsArray = JSArray()
        for (cookie in cookies) {
            val cookieJsPair = JSObject()
            cookieJsPair.put("key", cookie.name)
            cookieJsPair.put("value", cookie.value)
            cookiesJsArray.put(cookieJsPair)
        }
        val cookiesJsObject = JSObject()
        cookiesJsObject.put("cookies", cookiesJsArray)
        call.resolve(cookiesJsObject)
    }

    @PluginMethod
    fun getCookie(call: PluginCall) {
        val key = call.getString("key") ?: return call.reject("\"key\" must not be null")
        val cookie = cookieManager.getCookie(key)
        val cookieJsObject = JSObject()
        cookieJsObject.put("key", key)
        if (cookie != null) {
            cookieJsObject.put("value", cookie.value)
        } else {
            cookieJsObject.put("value", "")
        }
        call.resolve(cookieJsObject)
    }

    @PluginMethod
    fun deleteCookie(call: PluginCall) {
        val key = call.getString("key") ?: return call.reject("\"key\" must not be null")
        val url = getServerUrl(call) ?: return call.reject("Could not parse the server \"url\" value")

        cookieManager.deleteCookie(url, key)
        call.resolve()
    }

    @PluginMethod
    fun clearCookies(call: PluginCall) {
        cookieManager.deleteAllCookies()
        call.resolve()
    }
}